File: app/src/main/java/com/anatdx/nemuri/xposed/bridge/AppFreezeState.java
/* SPDX-License-Identifier: Apache-2.0 */
/*
 * Nemuri - Per-app foreground/background state for the auto-freeze engine.
 *
 * License: Apache-2.0
 *
 * Author: Anatdx
 */

package com.anatdx.nemuri.xposed.bridge;

import android.os.IBinder;

import androidx.annotation.NonNull;

import java.util.HashSet;
import java.util.Set;

// Tracks an app's visible activity tokens. visible flips only when the set goes empty<->non-empty;
// setVisible returns true only on an actual transition, so the engine schedules/cancels once.
final class AppFreezeState {
    private final Set<IBinder> visibleActivities = new HashSet<>();
    private boolean visible = false;

    synchronized boolean addActivity(@NonNull IBinder token) {
        visibleActivities.add(token);
        return setVisible(!visibleActivities.isEmpty());
    }

    synchronized boolean removeActivity(@NonNull IBinder token) {
        visibleActivities.remove(token);
        return setVisible(!visibleActivities.isEmpty());
    }

    synchronized boolean isVisible() {
        return visible;
    }

    private boolean setVisible(boolean value) {
        if (visible == value) {
            return false;
        }
        visible = value;
        return true;
    }
}

File: app/src/main/java/com/anatdx/nemuri/xposed/bridge/AppFreezeState.kt
/* SPDX-License-Identifier: Apache-2.0 */
/*
 * Nemuri - Per-app foreground/background state for the auto-freeze engine.
 *
 * License: Apache-2.0
 *
 * Author: Anatdx
 */

package com.anatdx.nemuri.xposed.bridge

import android.os.IBinder

// Tracks an app's visible activity tokens. visible flips only when the set goes empty<->non-empty;
// setVisible returns true only on an actual transition, so the engine schedules/cancels once.
class AppFreezeState {
    private val visibleActivities = HashSet<IBinder>()
    private var visible = false

    @Synchronized
    fun addActivity(token: IBinder): Boolean {
        visibleActivities.add(token)
        return setVisible(visibleActivities.isNotEmpty())
    }

    @Synchronized
    fun removeActivity(token: IBinder): Boolean {
        visibleActivities.remove(token)
        return setVisible(visibleActivities.isNotEmpty())
    }

    @Synchronized
    fun isVisible(): Boolean = visible

    private fun setVisible(value: Boolean): Boolean {
        if (visible == value) return false
        visible = value
        return true
    }
}
